include/cuco/static_map.cuh

```diff
@@ -179,6 +182,8 @@ class static_map {
                        empty_value<T> empty_value_sentinel,
                        KeyEqual const& pred                = {},
                        ProbingScheme const& probing_scheme = {},
+                       cuda_thread_scope<Scope> scope      = {},
+                       Storage storage                     = {},
```

Comment:
@sleeepyjack I just realized this might be overkill for host ctors since CTAD worked just fine.

In all use cases in libcudf, I have to add two additional empty brackets to adapt to this change. How can we improve this?

Reply:
The variadic ctor idea would come in handy here. Since we now control almost all argument types but Allocator and KeyEqual, we can give it another shot.

Apart from that we could reorder the arguments but that would be tedious and would require downstream projects to adapt that change.

I would not advise to remove them, because then CTAD might no longer work in certain cases IIRC.

Reply:
> Since we now control almost all argument types but Allocator and KeyEqual, we can give it another shot.

Allocator and KeyEqual are under control as well. So all arguments are of strong types.

> Apart from that we could reorder the arguments but that would be tedious and would require downstream projects to adapt that change.

The issue is that stream should be always the last argument but it is also the argument we custom the most...

Reply:
> Allocator and KeyEqual are under control as well.

Not sure. Both can be user-defined types at this point.

Allocator is easy, since it has to dulfill the allocator interface. We can simply use a trait for that one.

KeyEqual gave me some headaches in the past, since the only requirement is that it defines a binary operator for the comparison. So if we use that as a check, every argument type that has a binary call operator defined will be considered as a KeyEqual concept.

Reply:
> So if we use that as a check, every argument type that has a binary call operator defined will be considered as a KeyEqual concept.


It should be fine if the return type of the binary operator is bool and it's invokable with two arguments of the type `Key`

Reply:
> It should be fine if the return type of the binary operator is bool and it's invokable with two arguments of the type Key

That was also my initial thoughts on this.

Turns out everything that can be implicitly converted to a `bool` will suffice the first constraint.
The second constraint is also no longer valid, since we introduced the heterogeneous insert/lookup concept. We actually compare `ProbeKey`and `Key`, where `ProbeKey` type isn't known in the ctor.

I can think of two ways we can reliably detect a `KeyEqual`-like type:

1. We require the user to apply a wrapper or a tag to their user-defined `KeyEqual` and we than use that in a trait. It could be a wrapper like `cuco::key_equal<MyKeyEqual>` or we define a macro to do the wrapping for you, e.g. `CUCO__DECLARE_KEY_EQUAL(MyKeyEqual)`. Tags would also work, but I find them ugly from a design perspective. Then there's also the possibility of using inheritance, e.g., `struct MyKeyEqual : cuco::key_equal {…}`.
2. If we look at all the possible argument types, then there's only one other user-defined argument type (the `Allocator`) that _could possibly_ implement such a binary operator for whatever reason. So we can define the trait like this: `is_key_equal = is_binary_callable and not is_allocator_like`. There's one more thing required to make this actually work: Our probing scheme somehow also defines a binary callable operator. So the final trait would look like this: `is_key_equal = is_binary_callable and not is_allocator and not is_probing_scheme`.

Reply:
> Turns out everything that can be implicitly converted to a bool

I'm inclined to restrict the requirement here: it should be bool instead of "convertible to bool"

> We actually compare ProbeKey and Key, where ProbeKey type isn't known in the ctor.

Is it fair to say that supporting the comparison between `Key` and `ProbeKey` is a fancy extra and the `KeyEqual` being able to compare two `Key`s is a basic requirement?

Reply:
> I'm inclined to restrict the requirement here: it should be bool instead of "convertible to bool"

I totally agree. The problem was on the implementation side of the trait. I just couldn't think of a way to make this restriction work with the given STL tools. There might be a way.

> Is it fair to say that supporting the comparison between Key and ProbeKey is a fancy extra and the KeyEqual being able to compare two Keys is a basic requirement?

Fair point. I would say requiring a `Key<->Key` comparator (even it is actually never used) is reasonable.

Since we still need to check for '… and not is_allocator and not is_probing_scheme`, I would say proposal 2. plus the just discussed trait is the way to go here?